### PR TITLE
Moves affected by Terrain: Camouflage, Secret Power & Terrain Pulse

### PR DESCRIPTION
**Pull request by AsparagusEduardo**
## Description
- Camouflage's type affected by terrain.
- Secret Power's effect and animation affected by terrain. Addresses #1662 .
- Added new battle terrain defines for Secret Power and Camouflage's effects and types.
- Implemented Terrain Pulse. Addresses #1094 .
  - Without terrain:
  normal
  - Grassy Terrain:
  grassy
  - Misty Terrain:
  misty
  - Psychic Terrain:
  psychic
  - Electric Terrain
  electric

(Originally by Fontbane, resubmitted and modified with permission).

## **Discord contact info**
AsparagusEduardo#6051

### CODE REVIEW

**Comment by AsparagusEduardo** [comment]
I decided to leave the animation for later. Ready for review.

File: data/battle_anim_scripts.s [hunk 1]
```diff
@@ -13979,7 +13979,7 @@ Move_RISING_VOLTAGE::
 	end @to do:
 
 Move_TERRAIN_PULSE::
-	end @to do:
+	goto Move_DRAGON_PULSE @Temporary, needs to setup different animations based on terrain
```
Comment (ghoulslash) [inline data/battle_anim_scripts.s:13982]: What does the animation look like in game? A simple pulse animation that changes particle color based on terrain would be relatively easy to implement

File: data/battle_anim_scripts.s [hunk 2]
```diff
@@ -23251,16 +23251,21 @@ Move_SKY_UPPERCUT:
 	end
 
 Move_SECRET_POWER:
```
Comment (ghoulslash) [inline data/battle_anim_scripts.s:23253]: should these animations be updated to [Gen 7](https://bulbapedia.bulbagarden.net/wiki/Secret_Power_%28move%29#Generation_VII)?

**Review by ghoulslash — CHANGES_REQUESTED** [review]

**Comment by AsparagusEduardo** [comment]
Ready for re-review
- Added animations to Terrain Pulse:
  - Without terrain:
  normal
  - Grassy and Misty Terrains:
  grassy misty
  - Psychic and Terrains:
  psychic electric
  - Added config for Secret Power's effects and animations.
  - Added new battle terrains to complete Secret Power's, Nature Power's and Camouflage's data (plus added failsafes in case new terrains are added and are not accounted for these moves.

**Review by ghoulslash — APPROVED** [review]

**Pull request merged**